numba/listobject.py

```diff
@@ -0,0 +1,1089 @@
+        data_item = dm_item.as_data(builder, item)
+
+        ptr_item = cgutils.alloca_once_value(builder, data_item)
+
+        lp = _list_get_data(context, builder, tl, l)
+        status = builder.call(
+            fn,
+            [
+                lp,
+                _as_bytes(builder, ptr_item),
+            ],
+        )
+        return status
+
+    return sig, codegen
+
+
+@overload_method(types.ListType, 'append')
+def impl_append(l, item):
+    if not isinstance(l, types.ListType):
+        return
+
+    itemty = l.item_type
+
+    def impl(l, item):
+        casteditem = _cast(item, itemty)
+        status = _list_append(l, casteditem)
+        if status == ListStatus.LIST_OK:
+            return
+        elif status == ListStatus.LIST_ERR_NO_MEMORY:
+            raise MemoryError('Unable to allocate memory to append item')
+        else:
+            raise RuntimeError('list.append failed unexpectedly')
+
+    if l.is_precise():
+        # Handle the precise case.
+        return impl
+    else:
+        # Handle the imprecise case.
+        l = l.refine(item)
+        # Re-bind the item type match the arguments.
+        itemty = l.item_type
+        # Create the signature that we wanted this impl to have.
+        sig = typing.signature(types.void, l, itemty)
+        return sig, impl
+
+
+@register_jitable
+def handle_index(l, index):
+    """Handle index.
+
+    If the index is negative, convert it. If the index is out of range, raise
+    an IndexError.
+    """
+    # convert negative indices to positive ones
+    if index < 0:
+        index = len(l) + index
+    # check that the index is in range
+    if not (0 <= index < len(l)):
+        raise IndexError("list index out of range")
+    return index
+
+
+@register_jitable
+def handle_slice(l, s):
+    """Handle slice.
+
+    Convert a slice object for a given list into a range object that can be
+    used to index the list. Many subtle caveats here, especially if the step is
+    negative.
+    """
+    if len(l) == 0:  # ignore slice for empty list
+        return range(0)
+    ll, sa, so, se = len(l), s.start, s.stop, s.step
+    if se > 0:
+        start = max(ll + sa,  0) if s.start < 0 else min(ll, sa)
+        stop = max(ll + so, 0) if so < 0 else min(ll, so)
+    elif se < 0:
+        start = max(ll + sa,  -1) if s.start < 0 else min(ll - 1, sa)
+        stop = max(ll + so, -1) if so < 0 else min(ll, so)
+    else:
+        # should be caught earlier, but isn't, so we raise here
+        raise ValueError("slice step cannot be zero")
+    return range(start, stop, s.step)
+
+
+@intrinsic
+def _list_getitem(typingctx, l, index):
+    return _list_getitem_pop_helper(typingctx, l, index, 'getitem')
+
+
+@intrinsic
+def _list_pop(typingctx, l, index):
+    return _list_getitem_pop_helper(typingctx, l, index, 'pop')
+
+
+def _list_getitem_pop_helper(typingctx, l, index, op):
+    """Wrap numba_list_getitem and numba_list_pop
+
+    Returns 2-tuple of (intp, ?item_type)
```

Comment:
Yes, I copied this from the dict implementation without knowing exactly what it does. Probably worth removing this.

Reply:
I had a look at removing this, but I think I will need some help. The use of the optional in this case allows us to propagate all the way to the overload and then raise the error there. Otherwise, you have to raise from the intrinsic.